jira_sync/sync_mgr.py

```diff
@@ -0,0 +1,303 @@
+"""
+Manage synchronization between JIRA and upsteam git forges.
+"""
+
+import logging
+from collections.abc import Collection
+from functools import cached_property
+from typing import TYPE_CHECKING
+
+from jira import Issue as JiraIssue
+
+from .config import Config
+from .jira_wrapper import JIRA, JiraRunMode
+from .repositories import Instance, Issue, IssueStatus
+
+if TYPE_CHECKING:
+    from .config.model import Config, JiraConfig
+
+log = logging.getLogger(__name__)
+
+MatchedIssue = tuple[JiraIssue, Issue]
+
+
+class SyncManager:
+    """This class synchronizes tickets between git forges and JIRA."""
+
+    _config: "Config"
+    _jira_config: "JiraConfig"
+    _jira: JIRA
+    _instances_by_name: dict[str, Instance]
+
+    def __init__(self, config: Config, run_mode: JiraRunMode) -> None:
+        """Initialize a SyncManager instance.
+
+        :param config: A configuration structure
+        :param run_mode: Specify how to access JIRA (read-write, read-only
+            or dry-run)
+        """
+        self._config = config
+        self._jira_config = config.general.jira
+        self._jira_statuses = self._jira_config.statuses
+        self._jira_status_values = list(self._jira_statuses.model_dump().values())
+        self._jira = JIRA(self._jira_config, run_mode=run_mode)
+
+        self._instances_by_name = {
+            instance_name: Instance.from_config(
+                name=instance_name, config_path=config.config_path, config=instance_spec
+            )
+            for instance_name, instance_spec in config.instances.items()
+            if instance_spec.enabled
+        }
+
+    def sync_issues(self) -> None:
+        """Sync issues between the forges and JIRA.
+
+        This is an umbrella method, delegating to other methods for individual
+        tasks.
+        """
+        open_jira_issues = self.filter_open_jira_issues_by_forge_repo(
+            self.retrieve_open_jira_issues()
+        )
+        open_forge_issues = self.retrieve_open_forge_issues()
+
+        matched_issues, unmatched_jira_issues, unmatched_forge_issues = (
+            self.match_jira_forge_issues(open_jira_issues, open_forge_issues)
+        )
+        log.debug(
+            "=> %d matched, %d unmatched JIRA, %d unmatched forge issues",
+            len(matched_issues),
+            len(unmatched_jira_issues),
+            len(unmatched_forge_issues),
+        )
+
+        self.close_jira_issues(unmatched_jira_issues)
+
+        matched_issues |= self.create_or_reopen_jira_issues(unmatched_forge_issues)
+
+        self.reconcile_jira_forge_issues(matched_issues)
+        log.info("Done.")
```

Comment:
Would be nice to have a start of sync info message or say more about what is done.